consensus/src/consensus_observer/missing_blocks.rs

```diff
@@ -0,0 +1,703 @@
+
+        // Insert the block into the store using the round of the first block
+        match self
+            .blocks_missing_payloads
+            .lock()
+            .entry(first_block_epoch_round)
+        {
+            Entry::Occupied(_) => {
+                // The block is already in the store
+                warn!(
+                    LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                        "A missing block was already found for the given epoch and round: {:?}",
+                        first_block_epoch_round
+                    ))
+                );
+            },
+            Entry::Vacant(entry) => {
+                // Insert the block into the store
+                entry.insert(ordered_block);
+            },
+        }
+
+        // Perform garbage collection if the store is too large
+        self.garbage_collect_missing_blocks();
+    }
+
+    /// Garbage collects the missing blocks store by removing
+    /// the oldest blocks if the store is too large.
+    fn garbage_collect_missing_blocks(&self) {
+        // Calculate the number of blocks to remove
+        let mut blocks_missing_payloads = self.blocks_missing_payloads.lock();
+        let num_missing_blocks = blocks_missing_payloads.len() as u64;
+        let max_pending_blocks = self.consensus_observer_config.max_num_pending_blocks;
+        let num_blocks_to_remove = num_missing_blocks.saturating_sub(max_pending_blocks);
+
+        // Remove the oldest blocks if the store is too large
+        for _ in 0..num_blocks_to_remove {
+            if let Some((oldest_epoch_round, _)) = blocks_missing_payloads.pop_first() {
+                warn!(
+                    LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                        "The missing block store is too large: {:?} blocks. Removing the block for the oldest epoch and round: {:?}",
+                        num_missing_blocks, oldest_epoch_round
+                    ))
+                );
+            }
+        }
+    }
+
+    /// Removes and returns the block from the store that is now ready
+    /// to be processed (after the new payload has been received).
+    pub fn remove_ready_block(
+        &self,
+        received_payload_epoch: u64,
+        received_payload_round: Round,
+        block_payload_store: &BlockPayloadStore,
+    ) -> Option<OrderedBlock> {
+        // Calculate the round at which to split the blocks
+        let split_round = received_payload_round.saturating_add(1);
+
+        // Split the missing blocks at the epoch and round
+        let mut blocks_missing_payloads = self.blocks_missing_payloads.lock();
+        let mut blocks_at_higher_rounds =
+            blocks_missing_payloads.split_off(&(received_payload_epoch, split_round));
+
+        // Check if the last block is ready (this should be the only ready block).
+        // Any earlier blocks are considered out-of-date and will be dropped.
```

Comment:
what happens when there's out of order payload delivery? it looks like if we have block 1,2 and payload comes as 2 ,1 we'll only pop block 1 but not 2?

Reply:
Good question @zekun000 😄 Yeah, this isn't supported today. So, what will happen is in your example is:
1. `remove_ready_block` will return block 2 (and drop block 1)
2. We'll then try to process the ordered block for 2. But, it'll fail because it doesn't extend the last block ID (which is 0).
3. We'll then fallback to state sync on a commit message.

It probably makes sense to allow some support for more out-of-order messages, but I feel like this is a post-MVP thing. Thoughts?

Reply:
we probably should log something if this happens

Reply:
Done. Added a log if we drop any ready blocks in the pending store 😄